quipucords/api/aggregate_report/model.py

```diff
@@ -0,0 +1,291 @@
+    os_by_name_and_version: dict = field(default_factory=dict)
+    socket_pairs: int = 0
+    system_creation_date_average: date | None = None
+    vmware_hosts: int = 0
+    vmware_vm_to_host_ratio: float = 0
+    vmware_vms: int = 0
+
+    # Note: The following attributes come exclusively from raw facts.
+    # For now, that's our only option, and these are required outputs.
+    # TODO Refactor how we populate these when we restructure the underlying data.
+    missing_host_done: int = 0
+    openshift_cluster_instances: int = 0
+    openshift_node_instances: int = 0
+
+
+def get_aggregate_report_by_report_id(report_id: int) -> dict:
+    """
+    Get the aggregate report data for the given report ID.
+
+    TODO Turn this into a database lookup after we start storing AggregateReport.
+    """
+    report = Report.objects.get(pk=report_id)
+    return asdict(build_aggregate_report(report.id))
+
+
+def _aggregate_from_system_fingerprints(  # noqa: C901,PLR0912,PLR0915
+    aggregated: AggregateReport, fingerprints: Iterable[SystemFingerprint]
+) -> None:
+    """
+    Populate the AggregateReport object with data from the given SystemFingerprints.
+
+    TODO Try to break up some to this to reduce the noqa: C901,PLR0912,PLR0915.
+    """
+    os_by_name_and_version = defaultdict(lambda: defaultdict(int))
+    system_creation_dates = []
+    vmware_vms_by_host = defaultdict(list)
+
+    for fingerprint in fingerprints:
+        # Yes, we want *all* OS name/version combos regardless of `is_redhat`.
+        # This means output may include non-RHEL distros (e.g. Ubuntu, Debian).
+        # Note: Ansible sources are fingerprinted to store the Ansible controller
+        # version (e.g. "4.4.4") in SystemFingerprint.os_version, and that can
+        # results strange in output here like {"unknown": {"4.4.4": 1}}.
+        # See also: FingerprintTaskRunner._process_ansible_fact.
+        # Those versions may look weird, but Justin says to keep them all.
+        os_by_name_and_version[fingerprint.os_name or UNKNOWN][
+            fingerprint.os_version or UNKNOWN
+        ] += 1
+
+        source_types = [source.get("source_type") for source in fingerprint.sources]
+        has_network_source = DataSources.NETWORK in source_types
+        has_satellite_source = DataSources.SATELLITE in source_types
+        has_vcenter_source = DataSources.VCENTER in source_types
+        has_openshift_source = DataSources.OPENSHIFT in source_types
+
+        if not (fingerprint.is_redhat or has_openshift_source or has_satellite_source):
+            # For all other fingerprint facts we aggregate below here, we only want
+            # to count them if we positively identified Red Hat during fingerprinting.
+            # We treat everything found by OpenShift or Satellite to be "is_redhat=True"
+            # here even if the "is_redhat" attribute was not set on the fingerprint.
+            continue
```

Comment:
not something we can do about it in code, but we can get lots of false positives and false negatives with vcenter here.